pkg/storage/etcd3/etcdwatcher/watcher.go

```diff
@@ -0,0 +1,340 @@
+type watchChan struct {
+	watcher       *watcher
+	key           string
+	initialRev    int64
+	recursive     bool
+	filter        storage.FilterFunc
+	ctx           context.Context
+	cancel        context.CancelFunc
+	resultChan    chan watch.Event
+	eventChan     chan *event
+	resultQueue   *list.List
+	errChan       chan error
+	waitSync      chan struct{}
+	waitEventProc sync.WaitGroup
+}
+
+func newWatcher(client *clientv3.Client, codec runtime.Codec, versioner storage.Versioner) *watcher {
+	return &watcher{
+		client:    client,
+		codec:     codec,
+		versioner: versioner,
+	}
+}
+
+func (w *watcher) Watch(ctx context.Context, key string, rev int64, recursive bool, filter storage.FilterFunc) (watch.Interface, error) {
+	if recursive && !strings.HasSuffix(key, "/") {
+		key += "/"
+	}
+	wc := &watchChan{
+		watcher:     w,
+		key:         key,
+		initialRev:  rev,
+		recursive:   recursive,
+		filter:      filter,
+		resultChan:  make(chan watch.Event, 1),
+		eventChan:   make(chan *event, 100),
+		resultQueue: list.New(),
+		errChan:     make(chan error, 1),
+		waitSync:    make(chan struct{}),
+	}
+	wc.ctx, wc.cancel = context.WithCancel(ctx)
+	go wc.run()
+	wc.waitEventProc.Add(1)
+	return wc, nil
+}
+
+func (wc *watchChan) run() {
+	go wc.sync()
+	go wc.startWatching()
+	go wc.eventProcessor()
+
+	var errResult *watch.Event
+	select {
+	case err := <-wc.errChan:
+		errResult = wc.parseError(err)
+		wc.cancel()
+	case <-wc.ctx.Done():
+	}
+	// error result is guaranteed to be received by user before closing ResultChan.
+	if errResult != nil {
+		wc.resultChan <- *errResult
+	}
+	wc.waitEventProc.Wait()
+	close(wc.resultChan)
+}
+
+func (wc *watchChan) Stop() {
+	wc.cancel()
+}
+
+func (wc *watchChan) ResultChan() <-chan watch.Event {
+	return wc.resultChan
+}
+
+// sync tries to retrieve existing data send them to process.
+// The revision to watch will be set to the revision in response.
+func (wc *watchChan) sync() {
+	defer close(wc.waitSync)
+	if wc.initialRev != 0 {
+		return
+	}
+
+	opts := []clientv3.OpOption{}
+	if wc.recursive {
+		opts = append(opts, clientv3.WithPrefix())
+	}
+	getResp, err := wc.watcher.client.Get(wc.ctx, wc.key, opts...)
+	if err != nil {
+		go wc.sendError(err)
+		return
+	}
+	wc.initialRev = getResp.Header.Revision
+
+	for _, kv := range getResp.Kvs {
+		wc.sendEvent(parseEtcdKV(kv))
+	}
+}
+
+// startWatching does:
+// - wait for sync to complete
+// - watch on given key and send events to process.
+func (wc *watchChan) startWatching() {
+	select {
+	case <-wc.waitSync:
+	case <-wc.ctx.Done():
+	}
+	opts := []clientv3.OpOption{clientv3.WithRev(wc.initialRev + 1)}
+	if wc.recursive {
+		opts = append(opts, clientv3.WithPrefix())
+	}
+	wch := wc.watcher.client.Watch(wc.ctx, wc.key, opts...)
+	for wres := range wch {
+		if wres.Err() != nil {
+			// If there is an error on server (e.g. compaction), the channel will return it before closed.
+			go wc.sendError(wres.Err())
```

Comment:
Removing rev=0 is an API change.  Probably best not to try and get bogged
down with it in this PR.

On Mon, Apr 11, 2016 at 9:27 PM, Hongchao Deng <notifications@github.com>
wrote:

> In pkg/storage/etcd3/etcdwatcher/watcher.go
> <https://github.com/kubernetes/kubernetes/pull/23694#discussion_r59308994>
> :
>
> > +// - wait for sync to complete
> > +// - watch on given key and send events to process.
> > +func (wc *watchChan) startWatching() {
> > +	select {
> > +	case <-wc.waitSync:
> > +	case <-wc.ctx.Done():
> > +	}
> > +	opts := []clientv3.OpOption{clientv3.WithRev(wc.initialRev + 1)}
> > +	if wc.recursive {
> > +		opts = append(opts, clientv3.WithPrefix())
> > +	}
> > +	wch := wc.watcher.client.Watch(wc.ctx, wc.key, opts...)
> > +	for wres := range wch {
> > +		if wres.Err() != nil {
> > +			// If there is an error on server (e.g. compaction), the channel will return it before closed.
> > +			go wc.sendError(wres.Err())
>
> Let me think more on this. So I don't really like to idea to have any
> function blocked in select loop.
> That say, it's good to get rid of the goroutine here at startWatching().
> But not the one in transform().
> Let me know about your opinions.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23694/files/c421d9632816dbbc999a80fc9c80662d9495736c#r59308994>
>